pkg/apis/storage/validation/validation.go

```diff
@@ -266,3 +278,180 @@ func validateAllowedTopologies(topologies []api.TopologySelectorTerm, fldPath *f
 
 	return allErrs
 }
+
+// ValidateCSINodeInfo validates a CSINodeInfo.
+func ValidateCSINodeInfo(csiNodeInfo *storage.CSINodeInfo) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&csiNodeInfo.ObjectMeta, false, apivalidation.ValidateClassName, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateCSINodeInfoSpec(&csiNodeInfo.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, validateCSINodeInfoStatus(&csiNodeInfo.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+// ValidateCSINodeInfoUpdate validates a CSINodeInfo.
+func ValidateCSINodeInfoUpdate(new, old *storage.CSINodeInfo) field.ErrorList {
+	allErrs := ValidateCSINodeInfo(new)
+
+	// Spec is read-only
+	// If this ever relaxes in the future, make sure to increment the Generation number in PrepareForUpdate
+	if !apiequality.Semantic.DeepEqual(old.Spec, new.Spec) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), new.Spec, "field is immutable"))
+	}
+	return allErrs
+}
+
+// ValidateCSINodeInfoSpec tests that the specified CSINodeInfoSpec
+// has valid data.
+func validateCSINodeInfoSpec(
+	spec *storage.CSINodeInfoSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validateCSIDriverInfoSpecs(spec.Drivers, fldPath.Child("drivers"))...)
+	return allErrs
+}
+
+// ValidateCSINodeInfoStatus tests that the specified CSINodeInfoStatus
+// has valid data.
+func validateCSINodeInfoStatus(
+	status *storage.CSINodeInfoStatus, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validateCSIDriverInfoStatuses(status.Drivers, fldPath.Child("drivers"))...)
+	return allErrs
+}
+
+// ValidateCSIDriverInfoSpecs tests that the specified CSIDriverInfoSpecs
+// have valid data.
+func validateCSIDriverInfoSpecs(
+	drivers []storage.CSIDriverInfoSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if drivers == nil || len(drivers) == 0 {
+		return allErrs
+	}
+
+	driverNamesInSpecs := make(sets.String)
+	for i, driver := range drivers {
+		idxPath := fldPath.Index(i)
+		allErrs = append(allErrs, apivalidation.ValidateCSIDriverName(driver.Name, idxPath.Child("name"))...)
+		allErrs = append(allErrs, validateCSIDriverInfoSpecNodeID(driver.NodeID, idxPath.Child("nodeID"))...)
+		allErrs = append(allErrs, validateCSIDriverInfoSpecTopologyKeys(driver.TopologyKeys, idxPath.Child("topologyKeys"))...)
+
+		// check for duplicate entries for the same driver in specs
+		if driverNamesInSpecs.Has(driver.Name) {
+			allErrs = append(allErrs, field.Invalid(idxPath.Child("name"), driver.Name, fmt.Sprintf("duplicate entries found for driver: %s in driver specs", driver.Name)))
```

Comment:
Error messgaes for validation should be imperative statements.  In this case, you should use `field.Duplicate` rather than Invalid.

Reply:
Ok